🌐 translation page faq-state (#57)
reactjs/fr.reactjs.org: content/docs/faq-state.md

Ci-dessous quelques ressources utiles pour en apprendre plus à propos de quand utiliser `props` au lieu de `state`:
* [Props vs State](https://github.com/uberVU/react-guide/blob/master/props-vs-state.md)
* [ReactJS: Props vs. State](https://lucybain.com/blog/2016/react-state-vs-pros/)

### Pourquoi `setState` me retourne une mauvaise valeur ? {#why-is-setstate-giving-me-the-wrong-value}

Dans React, `this.props` et `this.state` représente l'un et l'autre les valeurs rendues, i.e. ce qui se trouve actuellement à l'écran.

Les appels à `setState` sont asynchrones—ne comptez pas sur `this.state` pour refléter la nouvelle valeur immédiatement après avoir appelé `setState`. Passez une fonction de mise à jour à la place d'un objet si vous devez calculer des valeurs en fonction de l'état actuel (voir ci-dessous pour plus de détails).

Exemple de code qui ne se comporte *pas* comme attendu :

```jsx
incrementCount() {
  // Note : cela ne va *pas* fonctionner comme prévu.
  this.setState({count: this.state.count + 1});
}

handleSomething() {
  // Disons que `this.state.count` commence à 0.
  this.incrementCount();
  this.incrementCount();
  this.incrementCount();

>   // Lorsque React refait le rendu du composant, `this.state.count` est à 1, pourtant, on espère avoir 3.

Reviewer: ```suggestion
  // Lorsque React rafraîchira le composant, `this.state.count` sera à 1, pourtant, on s'attendait à 3.
```

“When” en anglais est suivi du présent même pour un futur, il faut déceler le temps implicite 😉

  // C'est parce que la fonction `incrementCount()` ci-dessus lit `this.state.count`,

>   // mais React ne met pas à jour `this.state.count` tant que le composant n'a pas refait le rendu.

Reviewer: ```suggestion
  // mais React ne met pas à jour `this.state.count` tant que le composant n'est pas rafraîchi.
```

>   // Donc `incrementCount()` lit `this.state.count`, qui est égal à 0 à chaque fois, et le définit à 1.

Reviewer: ```suggestion
  // Du coup `incrementCount()` lit `this.state.count` qui est égal à 0 à chaque fois, et le définit à 1.
```

Author: Le fameux "du coup" Français !

>   // Donc `incrementCount()` finit par lire `this.state.count` à 0 à chaque fois et le définit à 1.

Reviewer 2: meme si la traduction litterale est bonne je trouve cela un peut confus en francias (finit par lire)
on ne pourrait pas mettre a la place:  Donc `incrementCount()` lit `this.state.count`, qui est egal a 0 a chaque fois, et le definit a 1

>   // La correctif est décrite ci-dessous !

Reviewer 2: Le correctif
décrit

}
```

Voir ci-dessous pour savoir comment résoudre ce problème.

### Comment mettre à jour l'état avec des valeurs qui dépendent de l'état actuel ? {#how-do-i-update-state-with-values-that-depend-on-the-current-state}

Passez une fonction au lieu d'un objet à `setState` pour vous assurer que l'appel utilise toujours la version la plus récente de l'état (voir ci-dessous).

> ### Qu'elle est la différence entre passer un objet ou une fonction dans `setState` ? {#what-is-the-difference-between-passing-an-object-or-a-function-in-setstate}

Reviewer: ```suggestion
### Quelle est la différence entre passer un objet ou une fonction dans `setState` ? {#what-is-the-difference-between-passing-an-object-or-a-function-in-setstate}
```

> Passer une fonction de mise à jour vous permet d'accéder à la valeur de l'état actuel à l'intérieur de la fonction de mise à jour. Comme les appels `setState` sont groupés en lot, cela vous permet d'enchaîner les mises à jour et de vous assurer qu'elles se réalisent les unes après les autres au lieu d'entrer en conflit :

Reviewer: ```suggestion
Passer une fonction de mise à jour vous permet d'accéder à la valeur à jour de l'état actuel au sein de cette fonction. Comme les appels `setState` sont groupés par lots, ça vous permet d'enchaîner les mises à jour et de vous assurer qu'elles sont effectuées les unes après les autres au lieu d'entrer en conflit :
```

> Passer une fonction de mise à jour vous permet d'accéder à la valeur de l'état actuel à l'intérieur de la fonction de mise à jour. Comme les appels `setState` sont groupés en lot, cela vous permet d'enchaîner les mises à jour et de vous assurer qu'elles se réalisent les unes par-dessus les autres au lieu d'entrer en conflit :

Reviewer 3: ```suggestion
Passer une fonction de mise à jour vous permet d'accéder à la valeur de l'état actuel à l'intérieur de la fonction de mise à jour. Comme les appels `setState` sont groupés en lot, cela vous permet d'enchaîner les mises à jour et de vous assurer qu'elles se réalisent les unes après les autres au lieu d'entrer en conflit :
```

Reviewer 3: De manière général, utiliser des indicateur géographique (_par-dessus_) plutôt que temporel (_après_) et toujours un peu délicat dans ce genre de contexte. Je pense qu'on peut se permettre cette légère divergence par rapport à la VO.

Dans le même esprit, il y a beaucoup de _ci-dessous_ dans le texte. C'est ok, mais j'aurai tendance à préférer _ci-après_ à la place.

Author: @JeremiePat Je pense que je suis encore influencée par mes (lointains) exercices de Maths avec "trouvez l'exemple ci-dessous" ou "ci-contre" :)

```jsx
incrementCount() {
  this.setState((state) => {

>     // Important: lire `state` au lieu de `this.state` lors de la mise à jour.

Reviewer: ```suggestion
    // Important : lisez `state` au lieu de `this.state` lors de la mise à jour.
```

    return {count: state.count + 1}
  });
}

handleSomething() {
  // Disons que `this.state.count` commence à 0.
  this.incrementCount();
  this.incrementCount();
  this.incrementCount();

>   // Si vous lisez `this.state.count` maintenant, cela serait toujours 0.

Reviewer: ```suggestion
  // Si vous lisiez `this.state.count` maintenant, il serait toujours à 0.
```

>   // Mais lorsque React refera le rendu du composant, cela sera 3.

Reviewer: ```suggestion
  // Mais quand React rafraîchira le composant, il vaudra bien 3.
```

}
```

> [En apprendre plus sur setState](/docs/react-component.html#setstate)

Reviewer: ```suggestion
[En apprendre davantage sur setState](/docs/react-component.html#setstate)
```

Trad normalisée de “learn more“

> ### Quand `setState` est-il asynchrone ? {#when-is-setstate-asynchronous}

Reviewer: ```suggestion
### Quand `setState` est-elle asynchrone ? {#when-is-setstate-asynchronous}
```

Actuellement, `setState` est asynchrone à l'intérieur des gestionnaires d'événements.

> Cela assure, par exemple, que si le `Parent` et l'`Enfant` appellent tous les deux `setState` lors d'un clic, l'`Enfant` ne refait pas le rendu deux fois. Au lieu de ça, React "efface" les mises à jour de l'état à la fin de l'événement du navigateur. Le résultat se traduit par une amélioration significative des performances pour les applications de grande ampleur.

Reviewer: ```suggestion
Ça permet de garantir, par exemple, que si `Parent` et `Child` appellent tous les deux `setState` lors d'un clic, `Child` ne sera pas rafraîchi deux fois. Au lieu de ça, React « apure » les mises à jour de l'état à la fin de l'événement du navigateur. Ça permet une amélioration significative des performances pour les applications de grande ampleur.
```

- On ne traduit pas les identifiants de code
- [Guillemets français](https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#guillemets-français)
- Glissement sémantique sur “flush”, qui désigne ici le fait de traiter d'un coup toute la file des mises à jour en attente.
- Trop de mot à mot 😉

> Cela assure, par exemple, que si le `Parent` et l'`Enfant` appellent tous les deux `setState` pendant un événement clic, l'`Enfant` ne refait pas le rendu deux fois. Au lieu de ça, React "efface" les mises à jour de l'état à la fin de l'événement du navigateur. Le résultat se traduit par une amélioration significative des performances pour les grandes applications.

Reviewer 2: évènement x2
grandes applications sonne bizarre. on peut pas plutot utiliser pour les applications lourdes / pour les applications importantes?

Author: Malheureusement `event` est bien présent deux fois dans le paragraphe original. Est ce que:

> Cela assure, par exemple, que si le `Parent` et l'`Enfant` appellent tous les deux `setState` lors d'un clic, l'`Enfant` ne refait pas le rendu deux fois. Au lieu de ça, React "efface" les mises à jour de l'état à la fin de l'événement du navigateur. Le résultat se traduit par une amélioration significative des performances pour les applications de grande ampleur.

te semble une meilleure traduction ?

> Il s'agit d'un détail de l'implémentation donc évitez de vous appuyer dessus. Dans des versions futures, React groupera les mises à jour par défaut dans de plus en plus de cas de figure.

Reviewer: ```suggestion
Il s'agit d'un détail d'implémentation donc évitez de vous appuyer dessus. Dans de futures versions, React groupera par défaut les mises à jour dans davantage encore de cas de figure.
```

> ### Pourquoi React ne met-il pas à jour `this.state` de manière synchrone ? {#why-doesnt-react-update-thisstate-synchronously}

Reviewer: ```suggestion
### Pourquoi React ne met-il pas à jour `this.state` de façon synchrone ? {#why-doesnt-react-update-thisstate-synchronously}
```

Je trouve que ça passe un peu mieux…?

> Comme expliqué dans la section précédente, React "attend" intentionnellement jusqu'à ce que tous les composants appellent `setState()` dans leurs gestionnaires d'événements avant de commencer à effectuer un nouveau rendu. Cela améliore la performance en évitant des rendus inutiles.

Reviewer: ```suggestion
Comme expliqué dans la section précédente, React « attend » volontairement que tous les composants aient fini d'appeler `setState()` dans leurs gestionnaires d'événements avant de commencer à mettre à jour les rendus. Ça améliore les performances en évitant des rafraîchissements inutiles.
```

> Cependant, vous vous demandez peut-être toujours pourquoi React ne met pas juste à jour `this.state` immédiatement sans faire de nouveau rendu.

Reviewer: ```suggestion
Cependant, vous vous demandez peut-être toujours pourquoi React ne met pas juste à jour `this.state` immédiatement sans rafraîchir.
```

Il y a deux raisons principales :

> * Cela briserait la cohérence entre `props` et `state`, entraînant des problèmes très difficiles à déboguer.

Reviewer: ```suggestion
* Ça briserait la cohérence entre `props` et `state`, entraînant des problèmes très difficiles à déboguer.
```

> * Cela rendrait certaines nouvelles fonctionnalités sur lesquelles nous travaillons impossibles à mettre en pratique.

Reviewer: ```suggestion
* Ça rendrait certaines nouvelles fonctionnalités sur lesquelles nous travaillons impossibles à implémenter.
```

Le terme est bien connu et beaucoup plus spécifique que « mettre en pratique » (qui se place plus du côté utilisateur qu'implémenteur d'une fonctionnalité)

> Ce [commentaire GitHub](https://github.com/facebook/react/issues/11527#issuecomment-360199710) plonge profondément dans les exemples spécifiques.

Reviewer: ```suggestion
Ce [commentaire GitHub](https://github.com/facebook/react/issues/11527#issuecomment-360199710) entre dans le détail d’exemples spécifiques.
```

Mot à mot…

### Devrais-je utiliser une bibliothèque de gestion d'état comme Redux ou Mobx ? {#should-i-use-a-state-management-library-like-redux-or-mobx}

> [Peut-être.](https://redux.js.org/faq/general#when-should-i-use-redux)

Reviewer: 😁

> C'est une bonne idée de bien connaître React avant d'ajouter des bibliothèques supplémentaires. Vous pouvez créer des applications assez complexes en utilisant uniquement React.

Reviewer: ```suggestion
Bien connaître React avant d'ajouter des bibliothèques supplémentaires reste une bonne idée. Vous pouvez créer des applications plutôt complexes en utilisant uniquement React.
```
